accounts/abi/bind/precompilebind/precompile_contract_test_template.go

```diff
@@ -0,0 +1,142 @@
+				{{- else}}
+				input, err := Pack{{$func.Normalized.Name}}()
+				{{- end}}
+				require.NoError(t, err)
+				return input
+			},
+			{{- if not $fail}}
+			// This test is for a successful call. You can set the expected output here.
+			// CUSTOM CODE STARTS HERE
+			ExpectedRes: []byte{},
+			{{- end}}
+			SuppliedGas: {{$func.Normalized.Name}}GasCost,
+			ReadOnly:    false,
+			ExpectedErr: {{if $fail}} ErrCannot{{$func.Normalized.Name}}.Error() {{- else}} "" {{- end}},
+		},
+		{{- end}}
+		{{- end}}
+		{{- if not $func.Original.IsConstant}}
+		"readOnly {{decapitalise $func.Normalized.Name}} should fail": {
+			Caller:	common.Address{1},
+			InputFn: func(t *testing.T) []byte {
+				{{- if len $func.Normalized.Inputs | lt 1}}
+				// CUSTOM CODE STARTS HERE
+				// populate test input here
+				testInput := {{capitalise $func.Normalized.Name}}Input{}
+				input, err := Pack{{$func.Normalized.Name}}(testInput)
+				{{- else if len $func.Normalized.Inputs | eq 1 }}
+				{{- $input := index $func.Normalized.Inputs 0}}
+				// CUSTOM CODE STARTS HERE
+				// set test input to a value here
+				var testInput {{bindtype $input.Type $structs}}
+				input, err := Pack{{$func.Normalized.Name}}(testInput)
+				{{- else}}
+				input, err := Pack{{$func.Normalized.Name}}()
+				{{- end}}
+				require.NoError(t, err)
+				return input
+			},
+			SuppliedGas:  {{$func.Normalized.Name}}GasCost,
+			ReadOnly:    true,
+			ExpectedErr: vmerrs.ErrWriteProtection.Error(),
+		},
+		{{- end}}
+		"insufficient gas for {{decapitalise $func.Normalized.Name}} should fail": {
+			Caller:	common.Address{1},
+			InputFn: func(t *testing.T) []byte {
+				{{- if len $func.Normalized.Inputs | lt 1}}
+				// CUSTOM CODE STARTS HERE
+				// populate test input here
+				testInput := {{capitalise $func.Normalized.Name}}Input{}
+				input, err := Pack{{$func.Normalized.Name}}(testInput)
+				{{- else if len $func.Normalized.Inputs | eq 1 }}
+				{{- $input := index $func.Normalized.Inputs 0}}
+				// CUSTOM CODE STARTS HERE
+				// set test input to a value here
+				var testInput {{bindtype $input.Type $structs}}
+				input, err := Pack{{$func.Normalized.Name}}(testInput)
+				{{- else}}
+				input, err := Pack{{$func.Normalized.Name}}()
+				{{- end}}
+				require.NoError(t, err)
+				return input
+			},
+			SuppliedGas: {{$func.Normalized.Name}}GasCost - 1,
+			ReadOnly:    false,
+			ExpectedErr: vmerrs.ErrOutOfGas.Error(),
+		},
+		{{- end}}
+	}
+	{{- if .Contract.AllowList}}
+	// Run tests with allow list tests.
+	// This adds allowlist run tests to your custom tests
+	// and runs them all together.
+	// Even if you don't add any custom tests, keep this. This will still
+	// run the default allowlist tests.
+	allowlist.RunPrecompileWithAllowListTests(t, Module, state.NewTestStateDB, tests)
+	{{- else}}
+	// Run tests.
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			test.Run(t, module, NewTestStateDB(t *testing.T))
```

Comment:
`NewTestStateDB(t *testing.T)` does not seem correct here.

Is it possible to include a test that generates a precompile then compiles or calls gofmt on the resulting code? We could do this as a separate PR or an issue, however I think it will make it easier to edit the templates in the long run.

Reply:
You mean like a unit test in the `precompile_bind_test`?